### ISSUE TITLE
First epoch runs very slowly

### ISSUE THREAD

**Issue by drbeane** [keras-team/keras#12561]
When training a CNN with Keras, the first epoch always takes substantially longer than subsequent epochs. I am wondering why, and if there is a way to fix this. 

I have seen similar questions asked elsewhere, including [this one ](https://github.com/keras-team/keras/issues/6503) that was answered by Francois Chollet. The answer Francois gave was that the first epoch took the same amount of time, but had a few additional seconds added to its timer to account for building the computational graph. That doesn't seem to be the issue in my case, as each batch in the first epoch takes about 100x longer to process than batches in later epochs. By the time we processing batches, the graph should be built, no?

As an example, I am currently attempting to used a pre-trained VGG16 model to build a classifier for the tiny-imagenet-200 dataset. Some relevant portions of my code are:

`conv_base = VGG16(weights='imagenet', include_top=False, input_shape=(64, 64, 3))`
`model = Sequential()`
`model.add(conv_base)`
`model.add(Flatten())`
`model.add(Dense(256, activation='relu'))`
`model.add(Dense(200, activation='softmax'))`
`conv_base.trainable = False`

`tr_size = 200 * 500`
`va_size = 200 * 50`
`bs = 256`

`tr_batches = kim.ImageDataGenerator().flow_from_directory(`
`    train_path, target_size=(64, 64), batch_size=bs)`

`va_batches = kim.ImageDataGenerator().flow_from_directory(`
`    valid_path, target_size=(64, 64), batch_size=bs)`

`tr_steps = math.ceil(tr_size / bs)`
`va_steps = math.ceil(va_size / bs)`

`opt = keras.optimizers.Adam(0.0001)`
`model.compile(loss='categorical_crossentropy', optimizer=opt, metrics=['accuracy'])`

`h = model.fit_generator(tr_batches, steps_per_epoch=tr_steps, epochs=30,`
`                        validation_data=va_batches, validation_steps=va_steps, verbose=1)`

Here is my progress for the first few epochs of training:

`Epoch 1/30`
`391/391 [==============================] - 30464s 78s/step - loss: 11.2031 - acc: 0.0355 - val_loss: 5.4949 - val_acc: 0.0290`
`Epoch 2/30
391/391 [==============================] - 299s 765ms/step - loss: 5.1959 - acc: 0.0446 - val_loss: 5.0930 - val_acc: 0.0560`
`Epoch 3/30
391/391 [==============================] - 297s 761ms/step - loss: 4.7879 - acc: 0.0943 - val_loss: 4.7841 - val_acc: 0.1067`
`Epoch 4/30
391/391 [==============================] - 296s 758ms/step - loss: 4.3646 - acc: 0.1489 - val_loss: 4.4764 - val_acc: 0.1402`
`Epoch 5/30
391/391 [==============================] - 297s 758ms/step - loss: 3.9655 - acc: 0.1966 - val_loss: 4.2195 - val_acc: 0.1743`

Any suggestions you can provide would be appreciated. Thank you.

**Comment by geshijoker** [comment]
This is the same problem to me

**Comment by ReinhardDominik** [comment]
According to #6503, the part of the graph which deals with training is build.

**Comment by mgroth0** [comment]
Are there any workarounds? I designed a script to test lots of different models, and the overwhelming bottleneck appears to be this "building of the computational graph" for each model while the subsequent time taken when additional epochs are added on per model is negligible. 

I'm wondering what approaches one might take to speed up this process. Is it generally faster to load the model from a file instead? Any other approaches?
